Add information about writing tests

ruishanteo opened the pull request:
Fixes #79

mfjkri commented on tutorials/ab3AddRemark.md:
````diff
@@ -52,6 +57,25 @@ public class RemarkCommand extends Command {
+- Since we created a new file RemarkCommand, we should also create a `RemarkCommandTest` file. JUnit tests are annotated with `@Test`. For now, we expect `execute` in `RemarkCommand` to fail. Thus, the test will reflect this expectation.
+
+- ```java
+    assertCommandFailure(new RemarkCommand(), model, MESSAGE_NOT_IMPLEMENTED_YET);
````
Should we include the full test instead? For clarity and consistency 
```java
@Test
 public void execute() {
     assertCommandFailure(new RemarkCommand(), model, MESSAGE_NOT_IMPLEMENTED_YET);
 }
 ```

ruishanteo replied:
Okay, done!

mfjkri commented on tutorials/ab3AddRemark.md:
```diff
@@ -141,6 +165,47 @@ public class RemarkCommand extends Command {
 
+<panel header="Update construction of `RemarkCommand` in `RemarkCommandTest`">
+
+- Since `RemarkCommand`’s constructor now accepts arguments, the tests will reflect that.
```
Since both the constructor and execute method of `RemarkCommand` was updated, how about phrasing this like this instead?

Since `RemarkCommand` now requires both an index and a remark for instantiation, the tests should be updated accordingly. Additionally, the execution of the command has been updated, so the tests should be adjusted to reflect these changes.

```java
@Test
public void execute() {
   final String remark = "Some remark";

   assertCommandFailure(new RemarkCommand(INDEX_FIRST_PERSON, remark), model,
           String.format(MESSAGE_ARGUMENTS, INDEX_FIRST_PERSON.getOneBased(), remark));
}
```

mfjkri reviewed (commented):
Hi Rui Shan! I believe this will be a great addition, as it allows students to follow along and learn about the various aspects of the code they need to write tests for as they go. I just have a few more additions for the tests based on the Git diffs in the tutorial.

Carlintyj commented:
Hello Rui Shan,

Great job on the proposed changes! I’ve reviewed your work, and I have a few additional suggestions that might help enhance the clarity and usefulness of the tutorial for the readers:

1. Explanation of @Test Annotation: Perhaps you could consider including a brief explanation of what the `@Test` annotation does when you first mention it. This will help readers who may not be as familiar with JUnit understand its purpose and how it fits into the testing process.

2. Context for Test-Writing: At the beginning of each test-writing section, it might be helpful to add a sentence or two explaining why the tests are being written at that point in the tutorial. This provides context and highlights the importance of testing in the development workflow.

3. Link to JUnit Resources: Including a link to the official JUnit documentation or a relevant tutorial could be beneficial for readers who want to explore unit testing in more depth.

Thank you for your hard work and besides these, LGTM!

damithc commented:
> 1. Explanation of @test Annotation: Perhaps you could consider including a brief explanation of what the `@Test` annotation does when you first mention it. This will help readers who may not be as familiar with JUnit understand its purpose and how it fits into the testing process.

@Carlintyj I'm not sure if we need this one. At this point students should have used JUnit in the iP already. We can point to the JUnit guide but we can't keep repeating all the basics in every tutorial :-)

Carlintyj commented:
Okay, sounds good to me to just point to the JUnit guide. 😃 Thanks prof @damithc

ruishanteo commented:
> 2. Context for Test-Writing: At the beginning of each test-writing section, it might be helpful to add a sentence or two explaining why the tests are being written at that point in the tutorial. This provides context and highlights the importance of testing in the development workflow.

Not sure what this refers to exactly. I have included short sentences/ clauses to explain why the tests are being written. For example: Since `RemarkCommand` now requires both an index and a remark for instantiation, the tests should be updated accordingly. Additionally, the execution of the command has been updated, so the tests should be adjusted to reflect these changes. Could you explain further/ give some examples on what this point refers to? Thanks!

> 3. Link to JUnit Resources: Including a link to the official JUnit documentation or a relevant tutorial could be beneficial for readers who want to explore unit testing in more depth.

Done! Added a link in the first panel.

Carlintyj commented:
> Since we created a new file RemarkCommand, we should also create a `RemarkCommandTest` file. JUnit tests are annotated with @Test. For now, we expect execute in RemarkCommand to fail. Thus, the test will reflect this expectation.

I was thinking instead of saying we should also create `RemarkCommandTest` file as such, we could perhaps just drop a sentence or two to give further clarification. For example: "it’s crucial to ensure that our code behaves as expected. This is why we write tests at each step of the process. These tests not only verify that the new functionality works correctly but also help prevent future changes from inadvertently breaking the code."

Hope this clarifies my intentions.

ruishanteo commented:
@Carlintyj 
I see! I added that explanation to the main panel explaining that tests are written. I think that the line given was general so I added it to the introductory panel instead of the individual panels where tests are explained.

jyue487 commented on tutorials/ab3AddRemark.md:
````diff
@@ -141,6 +168,53 @@ public class RemarkCommand extends Command {
+- ```java
+    @Test
+    public void execute() {
+    final String remark = "Some remark";
````
I am not too sure but should this be indented?

ruishanteo replied:
Thanks! Added the indentation.

jyue487 reviewed (commented):
Overall, LGTM! But are some of the contents overlapping with the Test section in IP?

damithc commented:
@ruishanteo Thanks for this PR.
I incorporated your changes in https://github.com/se-edu/guides/commit/a4cf2b5929667fd98f2f0f80c0755864eaefd8b4
